### ISSUE TITLE
On/off input for mover models

### ISSUE THREAD

**Issue by Mathadon** [lbl-srg/modelica-buildings#288]
Problem:
In practice pumps often have a constant pressure drop set point and they can be switched on or off remotely. Although the same behavior can be achieved by correctly setting dp_in, using dp_in is less convenient and intuitive.

Solution:
I propose the following changes:
1) add a boolean that enables you to activate an on/off boolean input signal to the pump
2) add a boolean that enables you to deactivate the dp_in input signal and use a constant dpSet instead.

For 1) the following model IDEAS.Fluid.Interfaces.OnOffInterface can be used:
```
partial model OnOffInterface 
  "Interface for either setting a device on using a parameter or using a realInput"
  parameter Boolean use_onOffSignal = false 
    "Set to true to switch device on/off using external signal";
  Modelica.Blocks.Interfaces.BooleanInput on if use_onOffSignal annotation (Placement(
        transformation(
        extent={{-20,-20},{20,20}},
        rotation=270,
        origin={-20,108})));
  parameter Boolean onOff=true "Set to true if device is on"
  annotation (Dialog(enable= not use_onOffSignal));
protected 
  Modelica.Blocks.Interfaces.BooleanInput on_internal 
    "Needed to connect to conditional connector";
equation 
  connect(on,on_internal);
  if not use_onOffSignal then
    on_internal=onOff;
  end if;
  annotation (Documentation(revisions="<html>
<ul>
<li>March 2014 by Filip Jorissen:<br/> 
Initial version
</li>
</ul>
</html>"));
end OnOffInterface;
```
For 2) a similar construction can be added. Due to the discrete on/off switching behavior it may be necessary to force the filteredInput to true when using the on/off input signal.

The same thing applies to the other pump types.

If you agree this is useful functionality I will propose an implementation?

Note that _in theory_ when having a constant dp_in it would be possible to transform the pump curve using similarity laws at initialization so that solving similarity laws at run time is no longer necessary. This may be useful on a longer term but can be seen separately from the changes above.

**Comment by Mathadon** [comment]
I would like to pick up this issue again. I can propose an implementation but we should decide on the following first: what happens if the pump is off? 
For instance consider the case where a ```FlowMachine_dp``` is turned off.
In practice neither ```dp``` or ```m_flow``` need to be zero since the pump will act as hydraulic resistance. I am however not sure what the flow characteristic of such a pump is like so it may not be easy to implement. Also the idea of ```FlowMachine_dp``` is to be a simple model so it is probably best to set m_flow = 0 or dp = 0 when the pump is off. How would you like to approach this?

**Comment by Mathadon** [comment]
@mwetter could you create a working branch so I can make a pull request?

**Comment by mwetter** [comment]
@Mathadon : I created the branch `issue288_onOffMovers` to review your proposed changes.

**Comment by mwetter** [comment]
@Mathadon : I looked at your code and think that a boolean input signal (on/off) is not the right data type. Many pumps have different stages. An on/off signal would not allow selecting the stage. I think it would be clearer to have an IntegerInput, and then by convention a value of 0 is off, 1 is the first stage, 2 the second stage, etc. 
Then, you can have a 
```
parameter Modelica.SIunits.Pressure head[:] "Pump had if use_stage = true"
annotation(enable=use_stages);
```
Note that `Buildings.Fluid.HeatExchangers.DXCoils.Examples.MultiStage` also uses an input signal called `stage`, so this would be consistent with the pumps.

You would need to make a new block that takes as a parameter the `head`, as input the integer signal `stage` and then outputs 
```
dp = if stage == 0 then 0 else head[stage];
```
 and writes an assertion if `stage < 0 or stage > size(head,1)`

Note that for consistency of the naming, please use `use_dp_in`, see also `Fluid.Sources.Boundary_pT`.

**Comment by Mathadon** [comment]
@mwetter I like your proposal. However I would prefer having both the stage and on/off inputs. In practice both inputs exist and it would be impractical when using the model to have to merge both inputs into one. Also the stage is often a fixed setpoint, whereas the on/off input is controlled using the building management system.
I hope adding both would not overcomplicate things..? Either way they should be disabled by default.

**Comment by mwetter** [comment]
@Mathadon I am not yet convinced that it would be impractical to simply set stage=0 if the pump should be off. For a pump controller with multiple stages, I would expect the stage to be reduced if the load gets smaller, and eventually `stage=0` is selected to switch the pump off. In my view there is no new functionality that would be enabled with the increased complexity of being able to select as an input `stage`, `on`, both or neither of the two? Hence I doubt that this increased complexity is useful for the typical user.

If you have a pump that is only on or off, you could use a controller that outputs either 0 (off) or 1 (on).

**Comment by mwetter** [comment]
@Mathadon 
I merged your version to the development branch.

I like the implementation but think a few items should be changed:

1. I don't understand why there is the option of a constant input. This seems overkill as I cannot think of a practical application in which the pumps or fans are never switched off. I suggest to remove this option.
2. The parameter that is now called `stageInputs` should be made protected (in the base class) and each mover should have a better named input, such as `heads`, `flow_rates`, `speeds` and `normalized_speeds`, whose value can then internally be assigned to `stateInputs`. This (a) allows to set the units so they show up in the Dymola GUI, (b) will be easier for the user to understand as we don't use the same name for different things, and (c) avoids that the value is retained if a user replaces the model from the Dymola GUI. Regarding (c), if you replace a pump with another one, then with the current implementation the value of `stageInputs` will not be removed, although it will be a wrong value.
3. I suggest to remove the instance `inputSwitch`. It is not a switch and not needed in my view.
4. The user's guide and the info section of each mover will also need to be updated, and we need to include examples or validations for each of the new option to make sure they are tested.
 
Please merge the latest version from the `Buildings` library to avoid future conflicts.

**Comment by Mathadon** [comment]
1) They don't need to be practical applications, for instance this is used already in some of the illustrative examples. Also in my model there do exist some pumps that are always on, with a fixed pressure head. I think it's better to provide this option since it's not really confusing either.
2) okay
3) inputSwitch makes it easier to connect 1) three possible inputs and 2) 2 possible outputs (filter or no filter). Removing it would require a lot of if-then-else connections, which are prone to errors and less easy to interpret. Therefore I'd like to propose to rename it block to ```inputPassThrough```
4) ok

I'll try to have a look at this in the coming days.

**Comment by Mathadon** [comment]
I made some of the changes. Do you agree with 1)?

**Comment by mwetter** [comment]
OK, let's do 1) with the default being so that old models still work.
You should also change the comment in `Use constant, real parameter input` because `constant`, `parameter` and `input` are all Modelica keywords. Using them here is confusing as an input cannot be a parameter and a constant cannot be a parameter.

Using the `inputPassThrough` is fine.

**Comment by Mathadon** [comment]
`Buildings.Fluid.Movers.Examples.MoverStages` contains following warning:
```
The initial conditions for variables of type Integer are not fully specified.
Dymola has selected default initial conditions.

Assuming fixed default start value for the discrete states:
floMacSta.extractor.index(start = 0)
```

The only way I can find to remove this is to duplicate `Modelica.Blocks.Routing.Extractor` and to add an initial condition: `pre(index)=0`. Adding this initial condition to the pump model is not possible since the component can be conditionally removed, resulting in a warning.

Do you see a different solution for this?

**Comment by mwetter** [comment]
@Mathadon 

I revised the model as follows:

I removed flow_rate and instead used m_flow_nominal as it does not make sense to use two different flow rates.

I renamed flow_rates to relativeMassFlowRates, and set in stageInput = m_flow_nominal*relativeMassFlowRates.
This allows scaling the size of the pump with m_flow_nominal, and it makes clear that flow rates are in mass and not volume flow rates.

Similarly, I removed head and used dp_nominal instead for FlowControlled_dp, and set stageInputs(each final unit="Pa") = dp_nominal*relativeHeads

See the updates I made to the user guide in `Buildings.Fluid.Movers` that describes these parameters.

I renamed Buildings.Fluid.Types.InputType.Stage to Buildings.Fluid.Types.InputType.Stages

What is still missing are more unit tests that cover all the input options for the two movers.

**Comment by Mathadon** [comment]
I had a look at the implementation. Some remarks:
1) I don't agree that the new implementation is more practical. A compromise could be to set the head default value to `heads = m_flow_nominal * {0.5,1}`. It can also be confusing since this is the only one of the three input types that uses a normalization. If you don't agree then I won't argue further.
2) `Buildings.Fluid.Movers.FlowControlled_dp` uses `dp_nominal` to normalise the stages but this parameter is 'hidden' in the `Dynamics` tab. I changed it to the 'Nominal conditions' group.
3)  `parameter Modelica.SIunits.Pressure[:] normalizedHeads` should be `parameter Real[:] normalizedHeads`. I fixed this.
4) The implementation is not consistent since pump_dp and pump_m_flow concatenate 0 to the stages input variable and pump_y and pump_Nrpm do not. I propose to remove the concatenation since there is no reason why the 0 'off' stage must exist. Moreover it offsets the indeces from what the user would expect based on the input he provided. I did not implement this yet.
5) Changed `%head` to `%dp_nominal` in `Buildings.Fluid.Movers.FlowControlled_dp` icon.
6) I added 2 new example models and unit tests and updated the `MoverStages` example and unit test. All models are now tested with all inputs. Note that `Buildings.Fluid.Movers.Examples.MoverContinuous` produces somewhat unexpected results for the mass flow rate as a function of the speed: 

![capture](https://cloud.githubusercontent.com/assets/5970879/9440266/f20eb6d6-4a6d-11e5-9f22-e7c2c8ae7772.PNG)
I was unable to run unit tests since I'm working on a backup machine without BuildingsPy.

@mwetter can you verify my changes and check 1) and 4) ?

**Comment by mwetter** [comment]
@Mathadon 

Regarding 4), I removed the concatenation in ffacf21 as you suggested because this is required for the models to be consistent.

Regarding 1), I changed in ed92d03 the heads and mass flow rates so that they are _not_ normalized anymore. I also updated the user guide accordingly.

All unit tests work. What remains is
- [ ] Integrate this with 457

**Comment by mwetter** [comment]
This is now merged into #457

**Comment by mwetter** [comment]
Closed as this is now on the master.
